### PR TITLE
linear operator

### PR DESCRIPTION
**Pull request by lambda7xx**
**Description of changes:**
- [x] update the linear operator
- [ ] `init_kernel` has some problem, we need to fix it. 

**Related Issues:**

Linked Issues:
- Issue #1071

Issues closed by this PR:
- Closes #

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1180)

### CODE REVIEW

File: lib/kernels/src/cuda/linear_kernels.cu [hunk 1]
```diff
@@ -13,76 +13,78 @@
...
-    case AC_MODE_TANH:
-      return true;
-    case AC_MODE_NONE:
-      return false;
-    default:
-      assert(0);
+// what's the float * one_ptr
+LinearPerDeviceState
+    init_kernel(PerDeviceFFHandle handle, Allocator allocator, float *one_ptr;
+                Activation activation,
```
Comment (reyna-abhyankar) [inline lib/kernels/src/cuda/linear_kernels.cu:28]: What is `Activation` I think this should be `ActiMode`

File: lib/kernels/src/cuda/linear_kernels.cu [hunk 2]
```diff
@@ -13,76 +13,78 @@
...
-    checkCUDNN(cudnnSetTensor4dDescriptor(m->outputTensor,
-                                          CUDNN_TENSOR_NCHW,
-                                          ff_to_cudnn_datatype(m->output_type),
-                                          batch_size,
-                                          channel,
-                                          1,
-                                          1));
-  }
+  // todo: how to use allocator to allocate memory for float * one_ptr, how many
+  // bytes to allocate?
```
Comment (reyna-abhyankar) [inline lib/kernels/src/cuda/linear_kernels.cu:71]: Use the same allocation method that is in the original constructor for `LinearPerDeviceState`

File: lib/runtime/src/ops/element_binary.cc [hunk 3]
```diff
@@ -213,7 +213,7 @@ OpTaskSignature init_signature<ELEMENTBINARY_INIT_TASK_ID>() {
 
   init.add_return_value<ElementBinaryPerDeviceState>();
 
-  return init;
+  return init; // todo:this may be wrong, because the headfile retrun void
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/element_binary.cc:216]: Don't worry about it in this PR. If it's a problem, you can open an issue

File: lib/runtime/src/ops/linear.cc [hunk 4]
```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
...
 using namespace FlexFlow::Kernels::Linear;
 
-static constexpr int KERNEL_IDX = 0;
-static constexpr int BIAS_IDX = 1;
+enum slots {
+  INPUT,
+  OUTPUT,
+  WEIGHT,
+  BIAS,
+  ATTR,
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/linear.cc:37]:
```suggestion
  ATTRS,
```

File: lib/runtime/src/ops/linear.cc [hunk 5]
```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
...
+  PROFILING,
+  HANDLE
+}; // Note: this needs add more
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/linear.cc:40]: Why?

File: lib/runtime/src/ops/linear.cc [hunk 6]
```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
...
+OpTaskInvocation init(LinearAttrs const &attrs) {
+  OpTaskBinding binding;
 
-  ParallelTensorShape input_shape = this->inputs[0]->get_shape();
-  ParallelTensorShape output_shape, kernel_shape, bias_shape;
-  LinearParams params = this->get_params();
-  params.construct_mappings(*this->parallel_dims_mapping, input_shape);
-  params.solve_dims(input_shape, output_shape, kernel_shape, bias_shape);
+  bind.bind_arg(HANDLE, ff_handle());
+  bind.bind_arg(ATTR, attrs);
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/linear.cc:46]:
```suggestion
  bind.bind_arg(ATTRS, attrs);
```

File: lib/runtime/src/ops/linear.cc [hunk 7]
```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
...
+  auto weight = acc.get_tensor<Permissions::RO>(WEIGHT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+  int out_dim = output.shape.at(ff_dim_t{0}) + 1;
+  int batch_size = output.shape.get_volume() / out_dim;
+
+  float *one_ptr;
+
+  DeviceSpecific<LinearPerDeviceState> state =
+      acc.create_device_specific<LinearPerDeviceState>(
+          init_kernel(handle,
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/linear.cc:92]: I don't think this function call matches what you have in the kernel.

File: lib/runtime/src/ops/linear.cc [hunk 8]
```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
...
+    init_task_impl(TaskArgumentAccessor const &acc) {
+  auto const &attrs = acc.get_argument<MultiHeadAttentionAttrs>(ATTRS);
+  Allocator allocator = acc.get_allocator();
+  PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
+
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  auto weight = acc.get_tensor<Permissions::RO>(WEIGHT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+  int out_dim = output.shape.at(ff_dim_t{0}) + 1;
+  int batch_size = output.shape.get_volume() / out_dim;
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/linear.cc:86]: We had agreed previously that `ff_dim_t{1}` is batch size. Is this equivalent?

File: lib/runtime/src/ops/linear.cc [hunk 9]
```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
...
+static DeviceSpecific<LinearPerDeviceState>
+    init_task_impl(TaskArgumentAccessor const &acc) {
+  auto const &attrs = acc.get_argument<MultiHeadAttentionAttrs>(ATTRS);
+  Allocator allocator = acc.get_allocator();
+  PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
+
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  auto weight = acc.get_tensor<Permissions::RO>(WEIGHT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+  int out_dim = output.shape.at(ff_dim_t{0}) + 1;
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/linear.cc:85]: Why the `+ 1`?

File: lib/runtime/src/ops/linear.cc [hunk 10]
```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
...
-  int output_n = sub_output.get_volume() / output_c;
-  LinearMeta *m = sim->linear_meta;
-  m->activation = activation;
-  m->input_type = inputs[0]->data_type;
-  m->weight_type = this->data_type;
-  m->output_type = outputs[0]->data_type;
-  assert(m->profiling == false);
+CostMetrics measure_operator_cost(SimEnvFactory const &sim_factory,
+                                  LinearAttrs const &attrs,
+                                  ParallelTensorShape const &input_shape,
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/linear.cc:202]: I think these should always be `InputParallelTensorDesc` instead of `ParallelTensorShape`

**Review by reyna-abhyankar — CHANGES_REQUESTED** [review]

File: deps/fmt
Comment (reyna-abhyankar) [inline deps/fmt:1]: Use most updated submodules

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/linear.cc` line 86 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379424984:-Npy5p8w5GMOgZ0LptEc:b2b56kx) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/linear.cc#L86)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

We had agreed previously that `ff_dim_t{1}` is batch size. Is this equivalent?
</blockquote></details>

ok, make sense

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/linear.cc` line 92 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379424690:-Npy6CkmBEUXtUrRknpg:bz68zmd) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/linear.cc#L92)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

I don't think this function call matches what you have in the kernel.
</blockquote></details>

where?  maybe this. should be 



_Code snippet:_
```cpp


  DeviceSpecific<LinearPerDeviceState> state =
      acc.create_device_specific<LinearPerDeviceState>(
          init_kernel(handle,
                      allocator,
                      one_ptr,
                      attrs.regularizer,
                      attrs.use_bias,
                      input.data_type,
                      weight.data_type,
                      output.data_type,
                      batch_size,
                      attrs.out_channels));
  return state;
}
```

File: lib/kernels/src/cuda/linear_kernels.cu [hunk 11]
```diff
@@ -13,76 +13,78 @@
...
+  checkCUDA(cudaMalloc(one_ptr, sizeof(float) * batch_size));
```
Comment (reyna-abhyankar) [inline lib/kernels/src/cuda/linear_kernels.cu:72]:
```suggestion
  checkCUDA(cudaMalloc(&one_ptr, sizeof(float) * batch_size));
```

**Review by lambda7xx — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1180)* status: 0 of 6 files reviewed, 12 unresolved discussions (waiting on @reyna-abhyankar)

___
*[`lib/kernels/src/cuda/linear_kernels.cu` line 28 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379422817:-Npy4I3n6ozFLoXDRffL:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/kernels/src/cuda/linear_kernels.cu#L28)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

What is `Activation` I think this should be `ActiMode`
</blockquote></details>

Done.
___
*[`lib/kernels/src/cuda/linear_kernels.cu` line 71 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379422965:-Npy5ByfAmVPedhcNQt0:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/kernels/src/cuda/linear_kernels.cu#L71)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Use the same allocation method that is in the original constructor for `LinearPerDeviceState`
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/element_binary.cc` line 216 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379423637:-Npy5Dp1-8AOz0MZCutt:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/element_binary.cc#L216)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Don't worry about it in this PR. If it's a problem, you can open an issue
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/linear.cc` line 37 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379423976:-Npy5ZC71LwOhDlVO1R2:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/linear.cc#L37)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

```suggestion
  ATTRS,
```
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/linear.cc` line 40 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379424007:-Npy5bBJEaTkVwGA1OS3:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/linear.cc#L40)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Why?
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/linear.cc` line 46 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379424134:-Npy5dtc-qybWiEi1DoI:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/linear.cc#L46)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

```suggestion
  bind.bind_arg(ATTRS, attrs);
```
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/linear.cc` line 85 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379425105:-Npy5hVj93xNk76oAM91:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/linear.cc#L85)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Why the `+ 1`? 
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/linear.cc` line 202 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1379427492:-Nq4sxCaCXxZPRxZ1W4x:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/lib/runtime/src/ops/linear.cc#L202)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

I think these should always be `InputParallelTensorDesc` instead of `ParallelTensorShape`
</blockquote></details>

Done.
___
*[`deps/fmt` line 0 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1180#gh-1456506559:-Nq4sy-eEBOzqGrluW2u:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/2f5908c89f260ad8ed6de8d024cbd318727167c8/deps/fmt#L0)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Use most updated submodules
</blockquote></details>

Done.

**Review by reyna-abhyankar — APPROVED** [review]

**Pull request merged**